### ISSUE TITLE
why minify more faster than esbuild?

### ISSUE THREAD

**Issue by qiudaoermu** [tdewolff/minify#588]
after testing on benchmark, minify is twice faster than esbuild. they both are written in Go.

**Comment by tdewolff** [comment]
You can compare minifiers here: https://github.com/privatenumber/minification-benchmarks. The reason that this minifier is so much faster is mainly due to a focus on high-performance code. It has many many performance considerations and optimizations so that it is as fast as possible, and additionally compresses better than ESBuild as well. For example, it only scans the input only once and doesn't have several passes such as other minifiers (which excludes a small set of minifications that can be applied). It reuses buffers and minifies in-place, avoiding new memory allocations. There is still room for improvement, such as using SIMD operations or by a goto-table instead of a switch statement in the lexer (only possible in ASM). But I want to avoid depending on the unsafe package or using difficult to maintain ASM code.

ESbuild to be fair has a different use-case and has some additional features that may get in the way of performance. It is more focused on a complete JS application and not a single file, and thus may get better results for compressing a complete JS package/module.

**Comment by qiudaoermu** [comment]
When comparing minifiers using minification-benchmarks, it is evident that Minify exhibits remarkable performance superiority over others when dealing with single files. However, how does it fare when confronted with multiple files?

**Comment by tdewolff** [comment]
In principle there is nothing special about multiple files. The library is built for single files, and you can process multiple files as you like (they are independent of each other). The command line tool uses a worker pool to process about 5 or 10 files concurrently, much faster than that is unlikely.

ESbuild by comparison, compresses modules/packages and by its nature these files are interdependent. This creates more minification opportunities, but you're unable to fully minify all files concurrently, part of it has to be serial. The minification benefits far outweigh the loss in performance, and ESbuild has gone out of their way to run as much as possible concurrently.

In short, if you want to minify a JS package or library (that you wrote), I would recommend ESbuild. If you have a few JS (or other) files, or you don't know what you're dealing with (for a client), use this library.
